### ISSUE TITLE
CA2000 false positive with async/await when method takes a potentially disposable parameter

### ISSUE THREAD

**Issue by jberezanski** [dotnet/roslyn-analyzers#2347]
#### Analyzer package

[Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers)

#### Package Version

v2.9.1

#### Diagnostic ID

[CA2000](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca2000-dispose-objects-before-losing-scope)

#### Repro steps

```csharp
using System;
using System.Threading.Tasks;

class C : IDisposable
{
    public void Dispose()
    {
    }

    public async Task<C> M1_Task(object context)
    {
        await Task.Yield();
        return new C();
    }

    public async Task M2_Task()
    {
        var c = await M1_Task(null).ConfigureAwait(false);
        c.Dispose();
    }
}
```

This is a slight variation of #2212. The inner method is async and takes a potentially disposable parameter.

#### Expected behavior

No CA2000 warnings.

#### Actual behavior

warning CA2000: Call System.IDisposable.Dispose on object created by 'M1_Task(null)' before all references to it are out of scope.

The warning appears when context is of type System.Object or a IDisposable reference type (such as C), but does not appear if context is of a type which does not implement IDisposable (other than System.Object).
